src/main/drivers/pwm_output_dshot_shared.h

```diff
@@ -0,0 +1,230 @@
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef USE_DSHOT
+
+#if defined(STM32F4) || defined(STM32F7)
+typedef DMA_Stream_TypeDef dmaStream_t;
+#else
+typedef DMA_Channel_TypeDef dmaStream_t;
+#endif
+
+FAST_RAM_ZERO_INIT static uint8_t dmaMotorTimerCount = 0;
+#ifdef STM32F7
+FAST_RAM_ZERO_INIT static motorDmaTimer_t dmaMotorTimers[MAX_DMA_TIMERS];
+FAST_RAM_ZERO_INIT static motorDmaOutput_t dmaMotors[MAX_SUPPORTED_MOTORS];
+#else
+static motorDmaTimer_t dmaMotorTimers[MAX_DMA_TIMERS];
+static motorDmaOutput_t dmaMotors[MAX_SUPPORTED_MOTORS];
+#endif
+
+#ifdef USE_DSHOT_TELEMETRY
+uint32_t readDoneCount;
+
+// TODO remove once debugging no longer needed
+FAST_RAM_ZERO_INIT uint32_t dshotInvalidPacketCount;
+FAST_RAM_ZERO_INIT uint32_t inputBuffer[DSHOT_TELEMETRY_INPUT_LEN];
+FAST_RAM_ZERO_INIT uint32_t setDirectionMicros;
+#endif
+
+motorDmaOutput_t *getMotorDmaOutput(uint8_t index)
```

Comment:
This is not a header file, as it does not contain only declarations and defines. Please make this into a separate compilation unit.

Reply:
I wanted to start sharing some code between the hal and non-hal implementations but still wanted most symbols to be static. I used the pattern of gyro_filter_impl.h here to share but still keep the implementation in one compilation unit for each target. If you think it's better to use separate compilation units I can do so.

Reply:
I would prefer to have separate compilation units - I don't think that sharing blocks of code that are not self-contained does much to improve maintainability. `gyro_filter_impl.h` (which should really be `gyro_filter_impl.c` as the contents of the file are C code, and not C headers) isn't a great pattern to follow - we went for this pattern to fix a particular problem after it had tripped us up a number of times, leading to hard to find bugs. The duality of stdperiph and hal code is something that goes through a lot of our driver layer, and it is generally implemented by having one shared compilation unit plus one for each framework, without sharing C code blocks, and it seems to work well and not cause the introduction of bugs.

Reply:
Ok. Thanks for the explanation.